board/opentrons/ot2/rootfs-overlay/bin/ot-machine-id-commit

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env sh
+
+# This script gives the system its persistent machine-id, as part of first-boot
+# setup.
+#
+# /etc/machine-id should contain a persistent, unique identifier for this
+# device. Userspace programs like journald rely on this.
+#
+# On our fresh SD card images, /etc/machine-id is unpopulated. When systemd sees
+# this, it autogenerates a transient machine-id and overlays it, so other
+# userspace programs can rely on its presence.
+#
+# +------------------------------------------+
+# | temp file, populated, created by systemd |
+# +------------------------------------------+
+#                      |
+#                 bind-mounted
+#                      |
+#                      V
+#           +---------------------+
+#           | regular file, empty |
+#           +---------------------+
+#                      |
+#                on filesystem
+#                      |
+#                      V
+#               /etc/machine-id
+#
+# Normally, systemd can, on its own, write back this autogenerated machine-id to
+# persistent storage. (See systemd-machine-id-commit.service.)
+# But in our case, that doesn't work because it expects our active rootfs to be
+# eventually writable, whereas we want to keep it always read-only.
+#
+# This script instead persists the current machine-id by:
+#
+# 1) Mounting our active rootfs in an additional, second place, off to the side
+#    where nothing can accidentally access it. And making that mount writable.
+# 2) Updating the underlying /etc/machine-id file through that second mount.
+#
+# Then, in future boots, the underlying /etc/machine-id file will be populated,
+# and systemd will reuse it instead of autogenerating a new one.
+#
+# See also the update-server service, which needs to apply a similar fixup to
+# system update images.
+
+
+set -eu -o pipefail
+
+MACHINE_ID_FILE=/etc/machine-id
+
+
+# todo(mm, 2021-02-25): Instead of checking like this, should we configure the
+# service to only run on systemd's notion of "first boot"? 
```

Comment:
probably not because then it won't run on update

Reply:
We don't need it to run on update, though, do we? update-server applies its own `/etc/machine-id` fixup before switching the active partition.

For comparison, this script, as currently implemented, also relies on that update-server fixup. It "runs" on the first boot after an update, but no-ops. It only does something meaningful on the first boot of a fresh SD card (modulo unexpected power loss events).

Reply:
Resolved in Slack.

> but then doesn't that mean that in-field updates won't do anything
>
> so any machine that already exists won't have this fix applied
> 
> this needs to happen once for every machine, right? So if it's onfirstboot, that would happen for machines with fresh flashes, but not for machines doing a root partition swap

> My understanding is that [this update-server code](https://github.com/Opentrons/opentrons/blob/304ce8df4f747788b1294fc7e90d60467569042d/update-server/otupdate/buildroot/update.py#L183-L184), introduced by [this PR](https://github.com/Opentrons/opentrons/pull/4058#pullrequest-317975055-permalink:~:text=By%20mounting%20the%20new%20rootfs%20in,with%20this%20commit.), takes care of that.
>
> * On any boot, no matter what we do, /etc/machine-id will already exist. It will either be a real file from us, or a random tmpfs file from systemd.
> * The update-server code linked above is in play for partition swaps. It takes the active rootfs’s /etc/machine-id, whether it’s real or tmpfs, and saves it to a real file in the standby rootfs, which it then switches the system to.
> * This implies that, with or without my PR #123, the problem fixes itself on OT-2s when they are first in-field-updated. (Well, it fixes itself going forward, at least. Older log files will still have their original random machine-ids.) From the first update onward, there will be a real persistent /etc/machine-id, and it will be stable.
> * My PR #123 is meant to cover everything before that first in-field update.

> ah. yeah that makes sense then